File: bikkuri/ngram.py
# Copyright (C) 2026 J. Nathanael Philipp (jnphilipp) <jnathanael@philipp.land>
#
# Bikkuri, calculate the surprisal of words in texts.
#
# This file is part of bikkuri.
#
# bikkuri is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# bikkuri is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with bikkuri. If not, see <http://www.gnu.org/licenses/>
"""N-gram surprisal calculations."""

import itertools
import math

from collections import Counter


class NGramSurprisal:
    """N-gram surprisal."""

    def __init__(self, n: int) -> None:
        """Init."""
        self._n = n
        self.__counter: Counter | None = None
        self.__suffix_counter: dict[str, Counter] | None = None

    def __call__(self, texts: list[list[str]]) -> list[list[tuple[str, float]]]:
        """Calculate the surprisal for the given texts."""
        return self.surprisal(texts)

    def _get_prefix(self, text: list[str], suffix: int) -> str | None:
        """Get the n-gram prefix of a word (suffix) in a text."""
        if self._n == 1:
            return None
        return "_".join(
            text[suffix - j] if suffix - j >= 0 else "_" for j in range(1, self._n)
        )

    def fit(self, texts: list[list[str]]) -> None:
        """Calculate the frequencies needed to calculate the n-gram surprisal."""
        if self.__counter is None:
            self.__counter = Counter()
        if self.__suffix_counter is None:
            self.__suffix_counter = {}

        for text in texts:
            for i, word in enumerate(text):
                self.__counter.update([word])
                if self._n > 1:
                    if word not in self.__suffix_counter:
                        self.__suffix_counter[word] = Counter()
                    self.__suffix_counter[word].update([self._get_prefix(text, i)])

    @property
    def n(self) -> int:
        """Get the type of n-gram."""
        return self._n

    def surprisal(self, texts: list[list[str]]) -> list[list[tuple[str, float]]]:
        """Calculate the surprisal for the given texts."""

        def get_prefix_count(prefix: str | None) -> int:
            assert self.__suffix_counter is not None

            if prefix is None:
                return 0

            return sum(
                v[prefix] for k, v in self.__suffix_counter.items() if prefix in v
            )

        def ngram_prob(text: list[str], word: str, idx: int) -> float:
            """Calculate the probability of a ngram."""
            assert self.__counter is not None
            assert self.__suffix_counter is not None

            prefix = self._get_prefix(text, idx)
            if word in self.__suffix_counter and prefix in self.__suffix_counter[word]:
                return self.__suffix_counter[word][prefix] / get_prefix_count(prefix)
            elif (
                word in self.__suffix_counter
                and prefix not in self.__suffix_counter[word]
            ):
                return self.__counter[word] / self.__counter.total()
            else:
                return 1.0 / (
                    get_prefix_count(prefix) + 1
                    if prefix
                    in itertools.chain(
                        *[
                            list(self.__suffix_counter[k].keys())
                            for k in self.__suffix_counter.keys()
                        ]
                    )
                    else self.__counter.total() + 1
                )

        assert self.__counter is not None

        if self._n == 1:
            return [
                [
                    (
                        word,
                        (
                            -1.0
                            * math.log2(
                                (self.__counter[word] / self.__counter.total())
                                if word in self.__counter
                                else (1.0 / (self.__counter.total() + 1.0))
                            )
                        ),
                    )
                    for word in text
                ]
                for text in texts
            ]
        else:
            return [
                [
                    (word, -1.0 * math.log2(ngram_prob(text, word, i)))
                    for i, word in enumerate(text)
                ]
                for text in texts
            ]


class UniGramSurprisal(NGramSurprisal):
    """Uni-gram surprisal."""

    def __init__(self) -> None:
        """Init."""
        super().__init__(1)


class BiGramSurprisal(NGramSurprisal):
    """Bi-gram surprisal."""

    def __init__(self) -> None:
        """Init."""
        super().__init__(2)

File: src/ngram.rs
// Copyright (C) 2026 J. Nathanael Philipp (jnphilipp) <jnathanael@philipp.land>
//
// Bikkuri, calculate the surprisal of words in texts.
//
// This file is part of bikkuri.
//
// bikkuri is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// bikkuri is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with bikkuri. If not, see <http://www.gnu.org/licenses/>
//
// SPDX-License-Identifier: GPL-3.0-or-later

use std::collections::{HashMap, VecDeque};

/// NGramSurprisal
pub struct NGramSurprisal {
    pub n: i64,
    counts: HashMap<String, i64>,
    suffix_counts: HashMap<String, HashMap<String, i64>>,
}

impl NGramSurprisal {
    pub fn new(n: i64) -> NGramSurprisal {
        NGramSurprisal {
            n: n,
            counts: HashMap::new(),
            suffix_counts: HashMap::new(),
        }
    }

    fn get_prefix<T, S: AsRef<str>>(&self, text: &T, suffix: i64) -> Option<String>
    where
        for<'a> &'a T: IntoIterator<Item = &'a S>,
    {
        if self.n <= 1 {
            return None;
        }

        let mut parts: VecDeque<String> = VecDeque::new();
        for _ in 1..self.n {
            parts.push_back("_".to_string());
        }
        let mut i: i64 = 0;
        for word in text {
            if i == suffix {
                break;
            } else {
                parts.pop_front();
                parts.push_back(word.as_ref().to_string());
            }
            i += 1;
        }

        Some(parts.into_iter().collect::<Vec<_>>().join("_"))
    }

    pub fn fit<T, U, S: AsRef<str>>(&mut self, texts: &T)
    where
        for<'a> &'a T: IntoIterator<Item = &'a U>,
        for<'a> &'a U: IntoIterator<Item = &'a S>,
    {
        for text in texts {
            let mut i: i64 = 0;
            for word in text {
                self.counts
                    .entry(word.as_ref().to_string())
                    .and_modify(|counter| *counter += 1)
                    .or_insert(1);
                if self.n > 1 {
                    let prefix: String = self.get_prefix(text, i).unwrap();
                    if let Some(x) = self.suffix_counts.get_mut(&word.as_ref().to_string()) {
                        x.entry(prefix)
                            .and_modify(|counter| *counter += 1)
                            .or_insert(1);
                    } else {
                        let mut suffix = HashMap::new();
                        suffix.insert(prefix, 1);
                        self.suffix_counts.insert(word.as_ref().to_string(), suffix);
                    }
                }
                i += 1;
            }
        }
    }

    pub fn surprisal<T, U, S: AsRef<str>>(&self, texts: &T) -> Option<Vec<Vec<(String, f64)>>>
    where
        for<'a> &'a T: IntoIterator<Item = &'a U>,
        for<'a> &'a U: IntoIterator<Item = &'a S>,
    {
        if self.counts.is_empty() {
            return None;
        }

        let prefix_count = |prefix: &String| -> f64 {
            self.suffix_counts
                .values()
                .map(|v| v.get(prefix).copied().unwrap_or(0) as f64)
                .sum()
        };

        let mut surprisal_texts = vec![];
        let total: f64 = self.counts.values().map(|&v| v as f64).sum();
        for text in texts {
            let mut i: i64 = 0;
            let mut surprisal_text = vec![];
            for word in text {
                let p: f64;
                let w = word.as_ref().to_string();
                if self.n == 1 {
                    if self.counts.contains_key(&w) {
                        p = *self.counts.get(&w).unwrap() as f64 / total;
                    } else {
                        p = 1.0 / (total + 1.0);
                    }
                } else {
                    let prefix: String = self.get_prefix(text, i).unwrap();
                    if self.suffix_counts.contains_key(&w)
                        && self.suffix_counts.get(&w).unwrap().contains_key(&prefix)
                    {
                        p = *self.suffix_counts.get(&w).unwrap().get(&prefix).unwrap() as f64
                            / prefix_count(&prefix);
                    } else if self.suffix_counts.contains_key(&w) {
                        p = *self.counts.get(&w).unwrap() as f64 / total;
                    } else if prefix_count(&prefix) == 0.0 {
                        p = 1.0 / (total + 1.0)
                    } else {
                        p = 1.0 / (prefix_count(&prefix) + 1.0);
                    }
                }
                surprisal_text.push((w, -p.log2()));
                i += 1;
            }
            surprisal_texts.push(surprisal_text);
        }

        Some(surprisal_texts)
    }
}
